zk/datastream/server/data_stream_server.go

```diff
@@ -206,78 +205,95 @@ func (srv *DataStreamServer) CreateStreamEntriesProto(
 	entryCount += len(gers)
 
 	var err error
-	if lastBatchNumber != batchNumber {
-		// we know we have some batch bookmarks to add, but we need to figure out how many because there
-		// could be empty batches in between blocks that could contain ger updates and we need to handle
-		// all of those scenarios
-		entryCount += int(3 * (batchNumber - lastBatchNumber)) // batch bookmark + batch start + batch end
-
-		// we don't write the batch end at the start of the stream when we're sending data to the executor
-		if mode == ExecutorOperationMode {
-			entryCount--
-		}
+
+	batchStart := batchNumber != lastBatchNumber
```

Comment:
So we aren’t sending empty batches to the DS even when they are closed?

Reply:
We send empty batches by comparing the previous batch with the latest.  So the stream population works on a block basis, so we ask what is batch for block X and what is the batch for block X-1, that will highlight a gap of empty batches which we pick up on later using the `batchGap` loop

Reply:
Allright, sure, so no blocks — no stream update.